0004-remove-spandata.md

```diff
@@ -0,0 +1,51 @@
+
+The second use case comes from the need to construct and report out of band spans, meaning that you're creating "custom" spans for an operation you don't own. One good example of this is a span sink that takes in structured logs that contain correlation IDs and a duration (e.g. from splunk) and [converts them](https://github.com/lightstep/splunktospan/blob/master/splunktospan/span.py#L43) to spans for your tracing system. [Another example](https://github.com/lightstep/haproxy_log2span/blob/master/lib/lib.go#L292) is running a sidecar on an HAProxy machine, tailing the request logs, and creating spans. SpanData allows you to report the out of band reporting case, whereas you can’t with the current Span API as you cannot set the start and end timestamp.
+
+I'd like to propose getting rid of SpanData and `tracer.recordSpanData()` and replacing it by allowing `tracer.startSpan()` to accept a start timestamp option and `span.finish()` to accept end timestamp option. This reduces the API surface, consolidating on a single span type. Options would meet the requirements for out of band reporting.
+
+## Internal details
+
+`startSpan()` would change so you can include an optional start timestamp and resource. When you have a span sink, out of band spans may have different resources than the tracer they are being reported to, so you want to pass an explicit resource. For `span.finish()` you would have an optional end timestamp. The exact implementation would be language specific, so some would use an options pattern with function overloading or variadic parameters, or add these options to the span builder.
+
+## Trade-offs and mitigations
+
+From https://github.com/open-telemetry/opentelemetry-specification/issues/71: If the underlying SDK automatically adds tags to spans such as thread-id, stacktrace, and cpu-usage when a span is started, they would be incorrect for out of band spans as the tracer would not know the difference between in and out of band spans. This can be mitigated by indicating that the span is out of band to prevent attaching incorrect information, possibly with an `isOutOfBand()` option.
+
+## Prior art and alternatives
+
+The OpenTracing specification for `tracer.startSpan()` includes an optional start timestamp and zero or more tags. It also calls out an optional end timestamp and bulk logging for `span.finish()`.
+
+## Open questions
+
+Imagine a scenario where you have a sidecar that accepts serialized spans from a legacy tracing system that you want to convert to OpenTelemetry spans. Those legacy spans have a trace and span ID you want to reuse. Should `startSpan()` include an option to explicitly set the span ID? Would it make more sense to take a direct dependency on the SDK's span type and translate to that?
```

Comment:
This is not just for legacy span cases. In the web tracing case we have a use for this as well. The browser does not send trace headers for the initial page navigation, so in order to relate the server and client spans for the initial load, we need the server to send back its parent trace and span ID, and have the browser client span (generated from browser performance timings) set its span ID to the provided parent span ID of the server. Here's [code in OpenCensus Web for this](https://github.com/census-instrumentation/opencensus-web/blob/4ab81112ce16216929a507b8e494a95d6d83298d/packages/opencensus-web-instrumentation-perf/src/initial-load-root-span.ts#L75).

It would be really helpful for `startSpan` to take an optional span ID. I'd rather not use `SpanData` because that's feels like extra complexity and in the case of the web browser, extra JS code that needs to be shipped and parsed and so would (slightly) slow down end user code the embeds the library.

Reply:
@bg451 would you be willing to update the language here to talk about the web use case in terms of receiving a span ID when creating a new span?